### PR TITLE
Reland Dispatch torch.norm to linalg.vector_norm and linalg.matrix_norm (#81761)

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84624

Reland https://github.com/pytorch/pytorch/pull/81761

Differential Revision: [D39332292](https://our.internmc.facebook.com/intern/diff/D39332292)

cc @VitalyFedyunin @ngimel @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @mcarilli @ptrblck @leslie-fang-intel @jgong5 @ezyang @fdrocha @peterbell10

### CODE REVIEW

**Comment by kit1980** [comment]
@kit1980 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39332292).

File: test/test_linalg.py [hunk 1]
```diff
@@ -1337,17 +1337,16 @@ def run_test_case(input, ord, dim, keepdim):
     def test_norm_fused_type_promotion(self, device, dtype):
         x = torch.randn(10, device=device, dtype=dtype)
 
-        def profile_and_check(fn, x, kwargs, fn_name):
+        def profile_and_check(fn, x, kwargs):
             with torch.profiler.profile(activities=(torch.profiler.ProfilerActivity.CPU,)) as p:
                 fn(x, **kwargs, dtype=torch.float)
             # smoke check that profiler returned some events
-            self.assertTrue(fn_name in map(lambda e: e.name, p.events()))
+            self.assertTrue("aten::linalg_vector_norm" in map(lambda e: e.name, p.events()))
```
Comment (kit1980) [inline test/test_linalg.py:1344]: Change to something like:
`self.assertTrue("aten::linalg_vector_norm" in (e.name for e in p.events()))`

(Flagged by flake8 in Meta-internal CI)

File: test/test_linalg.py [hunk 2]
```diff
@@ -1337,17 +1337,16 @@ def run_test_case(input, ord, dim, keepdim):
...
             # test that there was no explicit copy
             self.assertFalse("aten::to" in map(lambda e: e.name, p.events()))
```
Comment (kit1980) [inline test/test_linalg.py:1346]: Maybe also change while we are here.

`self.assertFalse("aten::to" in (e.name for e in p.events()))`

**Comment by kit1980** [comment]
I see two issues internally. One we know about already - https://github.com/pytorch/glow/issues/5994
Another one is in completely different project, trying to figure out what to do with it.

File: test/test_linalg.py (hunk 1 above)
Comment (Lezcano) [reply test/test_linalg.py]: @seemethere @malfet @peterbell10 it'd be cool to have this rule in OSS (avoid lambdas within maps) to avoid these reverts.

**Comment by Lezcano** [comment]
fixed those two!

File: test/test_linalg.py (hunk 1 above)
Comment (kit1980) [reply test/test_linalg.py]: I'm looking at bringing closer all Meta-internal and OSS linters. This check in particular should be easy to enable in OSS.

Comment (kit1980) [reply test/test_linalg.py]: Current OSS flake8 is older than Meta-internal flake8.
I have a draft PR to update https://github.com/pytorch/pytorch/pull/84732

**Comment by kit1980** [comment]
We have a fix for one of the internal issues, but this PR will need to be landed together with the fix - I'll handle that.
The glow issue still needs to be resolved.

**Comment by kit1980** [comment]
@lezcano The glow issue should be resolved for now. I'll update everything internally and try to land this one PR by landing the linked internal diff.
Hopefully there will be no ghstack-related issues. In general, ghsack complicates the things for me when internal changes are needed.

**Comment by lezcano** [comment]
Please ping me if you have any issues. If ghstack is a problem, I can pull this one out and we can land it on its own.

**Comment by kit1980** [comment]
@kit1980 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39332292).

**Review by kit1980 — APPROVED** [review]

**Comment by kit1980** [comment]
@lezcano looks like some functorch-related failures:

```
2022-09-24T00:03:27.7343497Z   File "test_testing.py", line 1815, in test_no_warning_on_import
2022-09-24T00:03:27.7343726Z     self.assertEquals(out, "")
2022-09-24T00:03:27.7344082Z AssertionError: '[W OperatorEntry.cpp:156] Warning: Overri[408 chars]l)\n' != ''
2022-09-24T00:03:27.7344556Z - [W OperatorEntry.cpp:156] Warning: Overriding a previously registered kernel for the same operator and the same dispatch key
2022-09-24T00:03:27.7345012Z -   operator: aten::linalg_matrix_norm
2022-09-24T00:03:27.7345244Z -     no debug info
2022-09-24T00:03:27.7345480Z -   dispatch key: FuncTorchBatched
```

**Comment by lezcano** [comment]
See how it goes now :)

**Comment by kit1980** [comment]
> See how it goes now :)

We have some hiccup here, I'll merge it next week.
Will need to rebase it again then.

**Comment by lezcano** [comment]
Sure thing. Just ping me if you need a more hands-on rebase :)

**Comment by lezcano** [comment]
@kit1980 what's the current state of this one?

**Comment by kit1980** [comment]
@lezcano sorry for delay. Can you resolve the merge conflict?

**Comment by lezcano** [comment]
Done!

**Comment by kit1980** [comment]
@kit1980 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39332292).

**Comment by kit1980** [comment]
@lezcano Now it's bunch of lint errors, looks like the merge gone wrong for common_methods_invocations.py?

**Comment by lezcano** [comment]
ugh, sorry for that. It should be fixed now.

**Comment by kit1980** [comment]
@kit1980 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39332292).

**Comment by kit1980** [comment]
Still something wrong with torch/testing/_internal/common_methods_invocations.py?

I got 

SyntaxError: closing parenthesis ')' does not match opening parenthesis '[' on line 7869
See https://www.flake8rules.com/rules/E999.html

**Comment by lezcano** [comment]
I think it should be fine really. See that almost all tests have already run and passed. Perhaps you are using an old version?

**Comment by kit1980** [comment]
@kit1980 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39332292).

**Comment by lezcano** [comment]
What's the state of this one @kit1980?

**Comment by kit1980** [comment]
> What's the state of this one @kit1980?

@lezcano sorry for delay. Look like a merge conflict again :-(

**Comment by lezcano** [comment]
@kit1980 rebased

**Comment by kit1980** [comment]
@kit1980 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39332292).

**Comment by kit1980** [comment]
@kit1980 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39332292).

**Pull request closed without merging**